alf/optimizers/optimizers.py

```diff
@@ -16,24 +16,75 @@
 import gin
 import torch
 
+import alf
 from alf.utils import common
+from alf.utils import tensor_utils
+from alf.optimizers.adam_tf import AdamTFUnwrapped
 
 
-def get_optimizer_with_empty_params(cls):
+def wrap_optimizer(cls):
     """A helper function to construct torch optimizers with
     params as [{'params': []}]. After construction, new parameter
     groups can be adde by using the add_param_group() method.
+
+    This wrapper also clips gradients first before calling ``step()``.
     """
     NewClsName = cls.__name__ + "_"
     NewCls = type(NewClsName, (cls, ), {})
+    NewCls.counter = 0
 
     @common.add_method(NewCls)
-    def __init__(self, **kwargs):
+    def __init__(self,
+                 gradient_clipping=None,
+                 clip_by_global_norm=False,
+                 name=None,
+                 **kwargs):
+        """
+        Args:
+            gradient_clipping (float): If not None, serve as a positive threshold
+            clip_by_global_norm (bool): If True, use `tensor_utils.clip_by_global_norm`
+                to clip gradient. If False, use `tensor_utils.clip_by_norms` for
+                each grad.
+            name (str): the name displayed when summarizing the gradient norm. If
+                None, then a global name in the format of "class_name_i" will be
+                created, where "i" is the global optimizer id.
+        """
         super(NewCls, self).__init__([{'params': []}], **kwargs)
+        self._gradient_clipping = gradient_clipping
+        self._clip_by_global_norm = clip_by_global_norm
+        self._name = name
```

Comment:
Is this name necessary?  Can use NewCls.__name__?

Reply:
If there are several Adam optimizers then their global norm summary will have conficts.

Reply:
You can still keep the counter.

Reply:
> You can still keep the counter.

Even the numbered names will be confusing like you don't know Adam_0 corresponds to your higher-level algorithm or your lower-level one. The order of creating optimizers can be uncertain or hard to find out.

Reply:
You are right. It's better to keep it if there are a lot of optimizers of the same type.